contracts/TokenHolder.sol

```diff
@@ -149,363 +171,347 @@ contract TokenHolder is MultiSigWallet {
-            uint8(0), // the amount in ether to be sent.
-            _data,
-            _nonce,
-            uint8(0), // gasPrice
-            uint8(0), // gasLimit
-            uint8(0), // gasToken
-            callPrefix, // 4 byte standard call prefix of the function to be called in the from contract.
-                        // This guarantees that a signed message can be only executed in a single instance.
-            uint8(0), // 0 for a standard call, 1 for a DelegateCall and 0 for a create opcode
-            '' // extraHash is always hashed at the end. This is done to increase future compatibility of the standard.
-        ));
+        key_ = ecrecover(_messageHash, _v, _r, _s);
     }
 
-    /**
-     * @notice private method to update ephemeralKeys mapping.
-     *
-     * @param _ephemeralKey Ephemeral Key which need to be added in ephemeralKeys mapping.
-     * @param _spendingLimit spending limit to be updated.
-     * @param _expirationHeight block height at which Ephemeral Key is expired.
-     */
-    function setEphemeralKeyData(
-        address _ephemeralKey,
-        uint256 _spendingLimit,
-        uint256 _expirationHeight
+    function getMessageHash (
+        bytes4 _callPrefix,
+        address _to,
+        bytes32 _dataHash,
+        uint256 _nonce
     )
         private
-        returns (bool)
+        view
+        returns (bytes32 messageHash_)
     {
-        ephemeralKeys[_ephemeralKey].spendingLimit = _spendingLimit;
-        ephemeralKeys[_ephemeralKey].nonce = 0;
-        ephemeralKeys[_ephemeralKey].expirationHeight = _expirationHeight;
-
-        return true;
+        messageHash_ = keccak256 (
+            abi.encodePacked  (
+                byte(0x19), // Starting a transaction with byte(0x19) ensure
+                            // the signed data from being a valid ethereum
+                            // transaction.
+                byte(0), // The version control byte.
+                address(this), // The from field will always be the contract
+                               // executing the code.
+                _to,
+                uint8(0), // The amount in ether to be sent.
+                _dataHash,
+                _nonce,
+                uint8(0), // gasPrice
+                uint8(0), // gasLimit
+                uint8(0), // gasToken
+                _callPrefix, // 4 byte standard call prefix of the
+                             // function to be called in the 'from' contract.
+                             // This guarantees that signed message can
+                             // be only executed in a single instance.
+                uint8(0), // 0 for a standard call, 1 for a DelegateCall and 2
+                          // for a create opcode
+                bytes32(0) // extraHash is always hashed at the end. This is
+                           // done to increase future compatibility of the
+                           // standard.
+            )
+        );
```

Comment:
Below are the differences I found from current working openst.js approach for signature construction and verification:

- In above we are hashing the _data in contract and pass it as parameter in getMessageHash. In  openst.js we don't hash the data when passing to getMessageHash. 
https://github.com/OpenSTFoundation/openst.js/blob/develop/utils/ExecutableTransaction.js

- Without hashing with ethereum standard prefix, signature verification was not working. We tried multiple ways to make it work with unsuccessful result. @deepesh-kn @sarvesh-ost also had to hash with prefix for signature verification.
https://github.com/OpenSTFoundation/openst-contracts/blob/token-economy-demo/contracts/TokenHolder.sol#L433
 
Considering above reasons please present modified signer verification with truffle test cases.  We spent quite a bit of time in openst.js making it work. So to accept above changes it will be great if we have tested with truffle.